src/openms/source/QC/DBSuitability.cpp

```diff
@@ -55,52 +65,387 @@ namespace OpenMS
+        continue;
+      }
+      if (decoy_1 < DBL_MAX && hit.getMetaValue("target_decoy") == "decoy")
+      {
+        decoy_2 = getRightScore_(hit);
+        break;
+      }
+    }
+
+    if (decoy_2 < DBL_MAX) // if there are two decoy hits
+    {
+      diff = abs(decoy_1 - decoy_2);
     }
 
+    // if there aren't two decoy hits DBL_MAX is returned
+    return diff;
+  }
+
+  double DBSuitability::getDecoyCutOff_(const vector<PeptideIdentification>& pep_ids, double reranking_cutoff_percentile) const
+  {
+    if (reranking_cutoff_percentile < 0 || reranking_cutoff_percentile > 1)
+    {
+      throw Exception::IllegalArgument(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "'reranking_cutoff_percentile' is not within its allowed range [0,1]. Please select a valid value.");
+    }
+
+    // get all decoy diffs of peptide ids with at least two decoy hits
+    vector<double> diffs;
+    for (const auto& pep_id : pep_ids)
+    {
+      double diff = getDecoyDiff_(pep_id);
+      if (diff < DBL_MAX)
+      {
+        diffs.push_back(diff);
+      }
+    }
+
+    if (double(diffs.size()) / pep_ids.size() < 0.2)
+    {
+      throw(Exception::MissingInformation(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "Under 20 % of peptide identifications have two decoy hits. This is not enough for re-ranking. Use the 'no_rerank' flag to still compute a suitability score."));
+    }
+
+    // sort the diffs decreasing and get the percentile one
+    UInt index = round(reranking_cutoff_percentile * diffs.size());
+    
+    if (index >= diffs.size())
+    {
+      return *max_element(diffs.begin(), diffs.end());
+    }
+
+    nth_element(diffs.begin(), diffs.begin() + index, diffs.end());
+
+    return diffs[index];
+  }
+
+  bool DBSuitability::isNovoHit_(const PeptideHit& hit) const
+  {
+    const boost::regex decoy_pattern(DecoyHelper::getPrefixRegex() + "|" + DecoyHelper::getSuffixRegex());
+    const set<String>& accessions = hit.extractProteinAccessionsSet();
+    for (const String& acc : accessions)
+    {
+      if (acc.find(Constants::UserParam::CONCAT_PEPTIDE) == String::npos && !boost::regex_search(acc, decoy_pattern))
+      {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  bool DBSuitability::passesFDR_(const PeptideHit& hit, double FDR) const
+  {
+    if (hit.getScore() > FDR) return false;
+    return true;
+  }
+
+  pair<String, Param> DBSuitability::extractSearchAdapterInfoFromMetaValues_(const ProteinIdentification::SearchParameters& search_params) const
+  {
     Param p;
+    // list of all allowed adapters
+    vector<String> working_adapters{ "CometAdapter", "CruxAdapter", "MSGFPlusAdapter", "MSFraggerAdapter", "MyriMatchAdapter", "OMSSAAdapter", "XTandemAdapter" };
+
+    vector<String> keys;
+    search_params.getKeys(keys);
+
+    // find adapter name
+    String adapter;
+    for (const String& key : keys)
+    {
+      for (const String& a : working_adapters)
+      {
+        if (key.compare(0, a.size(), a) == 0)
```

Comment:
a complicated way of saying `if (key == a)` ???

Reply:
ahh.. ok. you actually want the prefix, right?
Maybe annotate that...
and maybe also inlcude the ':' in the comparison?